### PR TITLE
Hierarchical Action List vel Component List.

### PR DESCRIPTION
**Pull request by farmaazon**
[ci no changelog needed]

### Pull Request Description

This PR introduces a new structure delivered by Searcher Controller: The Component List.

The Component list is hierarchical, and its structure reflects how the components are displayed in Component Browser - only "submodule" section at this point, other sections will be covered in next tasks.

This does not introduce anything visual; the structures are tested in unit tests.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide dist` and `./run ide watch`.

### CODE REVIEW

File: app/gui/controller/double-representation/src/module.rs [hunk 1]
```diff
@@ -282,6 +282,28 @@ impl QualifiedName {
             self.id.segments.pop();
         }
     }
+
+    /// Check if the name refers to some Library's top module.
```
Comment (wdanilo) [inline app/gui/controller/double-representation/src/module.rs]: Why library is written with upper-case? It is not a name in this context, is it?

File: app/gui/controller/double-representation/src/module.rs [hunk 2]
```diff
@@ -282,6 +282,28 @@ impl QualifiedName {
...
+    pub fn is_top_module(&self) -> bool {
+        self.id.segments.len() == 1
+    }
+
+    /// Get the top module containing the module referred by this name. Return self if it is already
+    /// a top module.
+    pub fn top_module(&self) -> Self {
+        Self {
+            project_name: self.project_name.clone(),
+            id:           Id { segments: self.id.segments.first().cloned().into_iter().collect() },
```
Comment (wdanilo) [inline app/gui/controller/double-representation/src/module.rs]: I believe this should be a method on ID.

File: app/gui/src/controller/searcher.rs [hunk 3]
```diff
@@ -930,25 +943,37 @@ impl Searcher {
...
+                    info!(this.logger, "Received suggestions from Language Server.");
+                    let list = this.make_action_list(responses.iter());
+                    let mut data = this.data.borrow_mut();
+                    data.actions = Actions::Loaded { list: Rc::new(list) };
+                    data.components = this.make_component_list(responses.iter());
+                }
+                Err(err) => {
+                    error!(
+                        this.logger,
+                        "Request for completions to the Language Server returned error: {err}"
```
Comment (wdanilo) [inline app/gui/src/controller/searcher.rs]: extract that to a var and make this 2 lines instead of 4 pls :)

File: app/gui/src/controller/searcher.rs [hunk 4]
```diff
@@ -1085,24 +1121,25 @@ impl Searcher {
     ///
     /// This is a workaround for Engine bug https://github.com/enso-org/enso/issues/1605.
     //TODO[ao] this is a temporary workaround.
-    fn add_enso_project_entries(libraries_cat_builder: &action::CategoryBuilder) -> FallibleResult {
+    fn add_enso_project_entries(libraries_cat_builder: &action::CategoryBuilder) {
```
Comment (wdanilo) [inline app/gui/src/controller/searcher.rs:1122]: remove "enso" name from the function pls. We should not encode the product name in the source code in other places than modules and crate names. I know this was not changed here, but as this code is touched, I added this comment for brevity :D

File: app/gui/src/controller/searcher.rs [hunk 5]
```diff
@@ -1085,24 +1121,25 @@ impl Searcher {
...
+        // We may unwrap here, because the constant is tested to be convertible to
+        // [`QualifiedName`].
+        let module = QualifiedName::from_text(ENSO_PROJECT_SPECIAL_MODULE).unwrap();
+        let self_type = tp::QualifiedName::from_text(ENSO_PROJECT_SPECIAL_MODULE).unwrap();
```
Comment (wdanilo) [inline app/gui/src/controller/searcher.rs:1126]: I wish this could be const expr. Maybe one day...

File: app/gui/src/controller/searcher.rs [hunk 6]
```diff
@@ -1592,14 +1665,34 @@ pub mod test {
...
+            });
+        searcher.reload_list();
+        test.run_until_stalled();
+        let components = searcher.components();
+        if let [module_group] = &components.top_modules()[..] {
+            assert_eq!(module_group.name, entry1.module.to_string());
+            let entries = module_group.entries.borrow();
+            assert_matches!(entries.as_slice(), [e1, e2] if e1.suggestion.name == entry1.name && e2.suggestion.name == entry9.name);
+        } else {
+            ipanic!("Wrong top modules in Component List: {components.top_modules():?}");
```
Comment (wdanilo) [inline app/gui/src/controller/searcher.rs:1690]: I would never stop reading this function as "I panic". Always when I see the name I'm feeling worried/disturbed :P

File: app/gui/src/controller/searcher/component/group.rs [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+/// A group of [`Component`]s.
+#[derive(Clone, CloneRef, Debug)]
+pub struct Group {
+    data: Rc<Data>,
+}
+
+impl Deref for Group {
+    type Target = Data;
+
```
Comment (wdanilo) [inline app/gui/src/controller/searcher/component/group.rs]: Minor thing: this nl is not needed.

File: app/gui/src/controller/searcher/component/group.rs [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+/// An immutable [`Group`] list, keeping the groups in alphabetical order.
+#[derive(Clone, CloneRef, Debug, Default)]
+pub struct List {
+    groups: Rc<Vec<Group>>,
+}
+
+impl Deref for List {
+    type Target = [Group];
+
```
Comment (wdanilo) [inline app/gui/src/controller/searcher/component/group.rs]: Minor thing: this nl is not needed.

**Review by wdanilo — APPROVED** [review]
Beautiful code and beautiful docs! Very understandable. It is very visible that you spent a lot of time cleaning and testing it. Looks gorgeous ❤️

File: app/gui/src/controller/searcher.rs (hunk 4 above)
Comment (farmaazon) [reply app/gui/src/controller/searcher.rs:1122]: The function name reflects the entity in our language. I will fill an issue for engine to remove Enso part from name.

File: app/gui/src/controller/searcher/component/builder.rs [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+        builder.extend(second_part);
+        let list = builder.build();
+
+        let top_modules: Vec<ComparableGroupData> =
+            list.top_modules.iter().map(Into::into).collect();
+        let expected = vec![
+            ComparableGroupData {
+                name:         "test.Test.TopModule1",
+                component_id: Some(0),
+                entries:      vec![2, 6, 3, 5],
```
Comment (akavel) [inline app/gui/src/controller/searcher/component/builder.rs]: This seems to not follow the order required by the Design Doc; am I missing something? Do we expect this sorting to be performed at some later stage?

I would expect the order to be:
`5, 6, 2, 3`

[design/design.md at main · enso-org/design](https://github.com/enso-org/design/blob/main/epics/component-browser/design.md#component-module-groups-during-searching)

> If user is not searching for components, each Component Module Group displays all of its items (alphabetically) and its sub-modules (alphabetically) in this order.

File: app/gui/src/controller/searcher/component/group.rs [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+            entry.module.name().into()
+        };
+        Self { data: Rc::new(Data::new_empty_visible(name, Some(component_id))) }
+    }
+
+    /// Check if group should be displayed in Component Browser.
+    pub fn check_visibility(&self) {
+        let entries = self.entries.borrow();
+        let filtered_out = entries.iter().filter(|c| c.is_filtered_out());
+        let visible = filtered_out.count() < entries.len();
```
Comment (akavel) [inline app/gui/src/controller/searcher/component/group.rs]: Would it make sense to use `Iterator::all` here?

File: app/gui/src/controller/searcher/component/builder.rs [hunk 11]
```diff
@@ -0,0 +1,339 @@
...
+        ];
+        assert_eq!(top_modules, expected);
+
+        let flattened_top_modules: Vec<ComparableGroupData> =
+            list.top_modules_flattened.iter().map(Into::into).collect();
+        let expected = vec![
+            ComparableGroupData {
+                name:         "test.Test.TopModule1",
+                component_id: Some(0),
+                entries:      vec![2, 6, 8, 9, 10, 3, 4, 5],
```
Comment (akavel) [inline app/gui/src/controller/searcher/component/builder.rs]: Is the order important here for us to assert? Per the design doc, the flattened modules seem to only become of interest once filtering is applied to the components list. The order when filtering is not applied seems an irrelevant implementation detail to me and allowed to change, so I wouldn't expect a test to fail if the code is changed such that the order is different here.

File: app/gui/src/controller/searcher/component/builder.rs [hunk 12]
```diff
@@ -0,0 +1,339 @@
...
+        let module_groups: BTreeMap<component::Id, ComparableGroupData> = list
+            .module_groups
+            .iter()
+            .map(|(id, m_groups)| (*id, (&m_groups.content).into()))
+            .collect();
+        let expected: BTreeMap<component::Id, ComparableGroupData> = [
+            (0, ComparableGroupData {
+                name:         "test.Test.TopModule1",
+                component_id: Some(0),
+                entries:      vec![2, 6, 3, 5],
```
Comment (akavel) [inline app/gui/src/controller/searcher/component/builder.rs]: Similar as above, shouldn't the order here follow the one required in the Design Doc?

**Review by akavel — CHANGES_REQUESTED** [review]
I don't see tests for search filtering, including sorting the results by match score; would you like to add such tests, or would you prefer that I add them?

**Review by akavel — APPROVED** [review]
